src/core/thread/osthread.d

```diff
@@ -2608,121 +2617,71 @@ else
+                asm { "stp x21, x22, %0" : "=m" regs[ 2], "=m" regs[3]; }
+                asm { "stp x23, x24, %0" : "=m" regs[ 4], "=m" regs[5]; }
+                asm { "stp x25, x26, %0" : "=m" regs[ 6], "=m" regs[7]; }
+                asm { "stp x27, x28, %0" : "=m" regs[ 8], "=m" regs[9]; }
+                asm { "str x29, %0"      : "=m" regs[10]; }
+                asm { "mov %0, sp"       : "=r" sp; }
             }
             else version (ARM)
             {
                 // Callee-save registers, according to AAPCS, section 5.1.1.
                 // arm and thumb2 instructions
                 size_t[8] regs = void;
-                __asm("stm  $0, {r4-r11}", "r", regs.ptr);
-                sp = __asm!(void*)("mov $0, sp", "=r");
+                asm { "stm %0, {r4-r11}" : : "r" regs.ptr : "memory"; }
+                asm { "mov %0, sp"       : "=r" sp; }
             }
-            else version (MIPS32)
-            {
-                // Callee-save registers, according to MIPS Calling Convention
-                size_t[8] regs = void;
-                __asm(`.set  noat;
-                       sw $$16, 0($0);
-                       sw $$17, 4($0);
-                       sw $$18, 8($0);
-                       sw $$19, 12($0);
-                       sw $$20, 16($0);
-                       sw $$21, 20($0);
-                       sw $$22, 24($0);
-                       sw $$23, 28($0);
-                       .set  at;`, "r", regs.ptr);
-                __asm(".set  noat; sw $$29, 0($0); .set  at;", "r", &sp);
-            }
-            else version (MIPS64)
+            else version (MIPS_Any)
             {
+                version (MIPS32)      enum store = "sw";
+                else version (MIPS64) enum store = "sd";
+                else static assert(0);
 
-                // Callee-save registers, according to MIPSpro N32 ABI Handbook,
-                // chapter 2, table 2-1.
+                // Callee-save registers, according to MIPS Calling Convention
+                // and MIPSpro N32 ABI Handbook, chapter 2, table 2-1.
                 // FIXME: Should $28 (gp) and $30 (s8) be saved, too?
                 size_t[8] regs = void;
-                __asm(`.set  noat;
-                       sd $$16,  0($0);
-                       sd $$17,  8($0);
-                       sd $$18, 16($0);
-                       sd $$19, 24($0);
-                       sd $$20, 32($0);
-                       sd $$21, 40($0);
-                       sd $$22, 48($0);
-                       sd $$23, 56($0);
-                       .set  at`, "r", regs.ptr);
-                __asm(".set  noat; sw $$29, 0($0); .set  at;", "r", &sp);
+                asm { ".set noat"; }
+                static foreach (i; 0 .. regs.length)
+                {{
+                    enum int j = 16 + i; // source register
+                    asm { (store ~ " $"~j.stringof~", %0") : "=m" regs[i]; }
```

Comment:
I've verified that these multiple asm statements with an indirect output operand each yield a final asm equivalent to the previous implementation (which didn't mark the `regs` memory as being clobbered).

Reply:
Btw, `{{` => separate scope for each static foreach iteration, required to prevent `enum int j` clashes. That enum is required to constant-fold the source register number, so that it can be converted to a string via `j.stringof`.